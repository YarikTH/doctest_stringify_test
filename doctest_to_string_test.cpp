#include "doctest/doctest.h"
#include "ApprovalTests.hpp"
#include <sstream>

#if __cplusplus >= 202002L
#   define CXX_STANDARD 20
#elif __cplusplus >= 201703L
#   define CXX_STANDARD 17
#elif __cplusplus >= 201402L
#   define CXX_STANDARD 14
#elif __cplusplus >= 201103L
#   define CXX_STANDARD 11
#endif

//#define USE_STL_STRINGIFIER
#ifdef USE_STL_STRINGIFIER
#   define DOCTEST_CPP CXX_STANDARD
#   define DOCTEST_STL_STRINGIFY_CHRONO
#   define DOCTEST_STL_STRINGIFY_FLIP
#   include "stl_stringifier.h"
#endif

#include <string>
#include <utility>
#include <tuple>
#include <array>
#include <valarray>
#include <initializer_list>
#include <vector>
#include <deque>
#include <list>
#include <forward_list>
#include <stack>
#include <queue>
#include <set>
#include <map>

#if CXX_STANDARD >= 14
#   include <string_view>
#endif

#if CXX_STANDARD >= 17
#   include <optional>
#   include <variant>
#endif

#if CXX_STANDARD >= 20
#   include <span>
#endif

using namespace ApprovalTests;

namespace
{

enum Enum
{
    E_VALUE = 1
};

enum EnumWithOss
{
    E_VALUE_WITH_OSS = 2
};

std::ostream& operator<<( std::ostream& os, EnumWithOss value )
{
    switch (value) {
    default:
        os << "???";
        break;
    case EnumWithOss::E_VALUE_WITH_OSS:
        os << "E_VALUE_WITH_OSS";
        break;
    }
    return os;
}

enum class EnumClass : int
{
    VALUE = 3
};

enum class EnumClassWithOss : int
{
    A = 4
};

std::ostream& operator<<( std::ostream& os, EnumClassWithOss value )
{
    switch (value) {
    default:
        os << "???";
        break;
    case EnumClassWithOss::A:
        os << "A";
        break;
    }
    return os;
}

struct Struct
{
};

struct StructOss
{
    int value;
};

inline std::ostream& operator<<(std::ostream& os, StructOss s)
{
    return os << s.value;
}

}

TEST_CASE("doctest::toString")
{
    std::ostringstream ss;
    
    auto formLabel = []( const char* str )
    {
        constexpr size_t LABEL_WIDTH = 30;
        const size_t len = strlen( str );
        const size_t spacesToAdd = (len < LABEL_WIDTH) ? LABEL_WIDTH - len : 0u;
        return std::string(4, ' ') + str + ':' + std::string(spacesToAdd, ' ') + ' ';
    };
    
#define DUMP_TYPE_VALUE(TYPE, VALUE) ss << formLabel( #TYPE ) << doctest::toString( static_cast<TYPE>(VALUE) ) << '\n'
    
    // https://en.cppreference.com/w/cpp/language/types
    ss << "Fundamental types:\n";
    
    // std::nullptr_t
    DUMP_TYPE_VALUE(std::nullptr_t, nullptr);
    
    // Signed and unsigned integer types
    DUMP_TYPE_VALUE(short int, -1);
    DUMP_TYPE_VALUE(unsigned short int, 1);
    DUMP_TYPE_VALUE(int, -2);
    DUMP_TYPE_VALUE(unsigned int, 2);
    DUMP_TYPE_VALUE(long int, -3);
    DUMP_TYPE_VALUE(unsigned long int, 3);
    DUMP_TYPE_VALUE(long long int, -4);
    DUMP_TYPE_VALUE(unsigned long long int, 4);
    
    // Boolean type
    DUMP_TYPE_VALUE(bool, true);
    
    // Character types
    DUMP_TYPE_VALUE(signed char, 'A');
    DUMP_TYPE_VALUE(unsigned char, 'B');
    DUMP_TYPE_VALUE(char, 'C');
    DUMP_TYPE_VALUE(wchar_t, L'D');
    DUMP_TYPE_VALUE(char16_t, L'E');
    DUMP_TYPE_VALUE(char32_t, L'F');
    
    // Floating-point types
    DUMP_TYPE_VALUE(float, 1.1f);
    DUMP_TYPE_VALUE(double, 2.2);
    DUMP_TYPE_VALUE(long double, 3.3);
    
    #define DUMP_POINTER_TYPE_VALUE(TYPE, VALUE) ss << formLabel( #TYPE "*" ) << doctest::toString( reinterpret_cast<TYPE*>(VALUE) ) << '\n' // NOLINT
    
    // Signed and unsigned integer types
    DUMP_POINTER_TYPE_VALUE(short int, 0);
    DUMP_POINTER_TYPE_VALUE(unsigned short int, 2);
    DUMP_POINTER_TYPE_VALUE(int, 3);
    DUMP_POINTER_TYPE_VALUE(unsigned int, 4);
    DUMP_POINTER_TYPE_VALUE(long int, 5);
    DUMP_POINTER_TYPE_VALUE(unsigned long int, 6);
    DUMP_POINTER_TYPE_VALUE(long long int, 7);
    DUMP_POINTER_TYPE_VALUE(unsigned long long int, 8);
    
    // Boolean type
    DUMP_POINTER_TYPE_VALUE(bool, 9);
    
    // Character types
    DUMP_POINTER_TYPE_VALUE(signed char, 10);
    DUMP_POINTER_TYPE_VALUE(unsigned char, 11);
    DUMP_POINTER_TYPE_VALUE(char, 12);
    DUMP_POINTER_TYPE_VALUE(wchar_t, 13);
    DUMP_POINTER_TYPE_VALUE(char16_t, 14);
    DUMP_POINTER_TYPE_VALUE(char32_t, 15);
    
    // Floating-point types
    DUMP_POINTER_TYPE_VALUE(float, 16);
    DUMP_POINTER_TYPE_VALUE(double, 17);
    DUMP_POINTER_TYPE_VALUE(long double, 18);
    
    #define DUMP_CONST_POINTER_TYPE_VALUE(TYPE, VALUE) ss << formLabel( "const " #TYPE "*" ) << doctest::toString( reinterpret_cast<const TYPE*>(VALUE) ) << '\n' // NOLINT
    
    // Signed and unsigned integer types
    DUMP_CONST_POINTER_TYPE_VALUE(short int, 0);
    DUMP_CONST_POINTER_TYPE_VALUE(unsigned short int, 2);
    DUMP_CONST_POINTER_TYPE_VALUE(int, 3);
    DUMP_CONST_POINTER_TYPE_VALUE(unsigned int, 4);
    DUMP_CONST_POINTER_TYPE_VALUE(long int, 5);
    DUMP_CONST_POINTER_TYPE_VALUE(unsigned long int, 6);
    DUMP_CONST_POINTER_TYPE_VALUE(long long int, 7);
    DUMP_CONST_POINTER_TYPE_VALUE(unsigned long long int, 8);
    
    // Boolean type
    DUMP_CONST_POINTER_TYPE_VALUE(bool, 9);
    
    // Character types
    DUMP_CONST_POINTER_TYPE_VALUE(signed char, 10);
    DUMP_CONST_POINTER_TYPE_VALUE(unsigned char, 11);
    DUMP_CONST_POINTER_TYPE_VALUE(char, 12);
    DUMP_CONST_POINTER_TYPE_VALUE(wchar_t, 13);
    DUMP_CONST_POINTER_TYPE_VALUE(char16_t, 14);
    DUMP_CONST_POINTER_TYPE_VALUE(char32_t, 15);
    
    // Floating-point types
    DUMP_CONST_POINTER_TYPE_VALUE(float, 16);
    DUMP_CONST_POINTER_TYPE_VALUE(double, 17);
    DUMP_CONST_POINTER_TYPE_VALUE(long double, 18);
    
    ss << "ะก arrays:\n";
    
    {
        const char cArray[] = {'f', 'o', 'o'};
        ss << formLabel( "const char[]" ) << doctest::toString( cArray ) << '\n';
    }
    
    {
        char cArray[] = {'b', 'a', 'r'};
        ss << formLabel( "char[]" ) << doctest::toString( cArray ) << '\n';
    }
    
    {
        const int cArray[] = {1, 2, 3};
        ss << formLabel( "const int[]" ) << doctest::toString( cArray ) << '\n';
    }
    
    {
        int cArray[] = {4, 5, 6};
        ss << formLabel( "int[]" ) << doctest::toString( cArray ) << '\n';
    }
    
    ss << "enumerations:\n";
    DUMP_TYPE_VALUE(Enum, Enum::E_VALUE);
    DUMP_TYPE_VALUE(EnumWithOss, EnumWithOss::E_VALUE_WITH_OSS);
    DUMP_TYPE_VALUE(EnumClass, EnumClass::VALUE);
    DUMP_TYPE_VALUE(EnumClassWithOss, EnumClassWithOss::A);
    
    ss << "structs:\n";
    
    {
        Struct s;
        ss << formLabel( "struct" ) << doctest::toString( s ) << '\n';
    }
    
    {
        StructOss s {};
        s.value = 777;
        ss << formLabel( "struct with oss" ) << doctest::toString( s ) << '\n';
    }

    ss << "std container and classes (c++11):\n";

    {
        const std::string value = "Hello world!";
        ss << formLabel( "std::string" ) << doctest::toString( value ) << '\n';
    }

    {
        std::pair<int, float> value { 1, 2.3f };
        ss << formLabel( "std::pair" ) << doctest::toString( value ) << '\n';
    }

    {
        std::tuple<bool, int, double> value { false, 7, 6.6 };
        ss << formLabel( "std::tuple" ) << doctest::toString( value ) << '\n';
    }

    {
        std::array<int, 3> value { 1, 5, 8 };
        ss << formLabel( "std::array" ) << doctest::toString( value ) << '\n';
    }

    {
        std::valarray<int> value { 7, 3, 2 };
        ss << formLabel( "std::valarray" ) << doctest::toString( value ) << '\n';
    }

    {
        std::initializer_list<int> value { 1, 7, 0 };
        ss << formLabel( "std::initializer_list" ) << doctest::toString( value ) << '\n';
    }

    {
        std::vector<int> value { 4, 2, 7 };
        ss << formLabel( "std::vector" ) << doctest::toString( value ) << '\n';
    }

    {
        std::deque<int> value { 1, 1, 6, 2 };
        ss << formLabel( "std::deque" ) << doctest::toString( value ) << '\n';
    }

    {
        std::list<int> value { 7, 5, 2, 1 };
        ss << formLabel( "std::list" ) << doctest::toString( value ) << '\n';
    }

    {
        std::forward_list<int> value { 7, 0, 3, 7, 4 };
        ss << formLabel( "std::forward_list" ) << doctest::toString( value ) << '\n';
    }

    {
        std::stack<int> value {{ 1, 2, 3, 4, 5 }};
        ss << formLabel( "std::stack" ) << doctest::toString( value ) << '\n';
    }

    {
        std::queue<int> value {{ 1, 2, 3, 4, 5 }};
        ss << formLabel( "std::queue" ) << doctest::toString( value ) << '\n';
    }

    {
        const auto data = { 1, 8, 5, 6, 3 };

        std::priority_queue<int> value { data.begin(), data.end() };
        ss << formLabel( "std::priority_queue" ) << doctest::toString( value ) << '\n';
    }
    
    {
        std::set<int> value {{ 1, 2, 2, 6, 1, 4 }};
        ss << formLabel( "std::set" ) << doctest::toString( value ) << '\n';
    }
    
    {
        std::multiset<int> value {{ 1, 2, 2, 6, 1, 4 }};
        ss << formLabel( "std::multiset" ) << doctest::toString( value ) << '\n';
    }
    
    {
        std::map<int, std::string> value {{ { 1, "foo" }, { 2, "bar" }, { 1, "baz" } }};
        ss << formLabel( "std::map" ) << doctest::toString( value ) << '\n';
    }
    
    {
        std::multimap<int, std::string> value {{ { 1, "foo" }, { 2, "bar" }, { 1, "baz" } }};
        ss << formLabel( "std::multimap" ) << doctest::toString( value ) << '\n';
    }

#if CXX_STANDARD >= 14
    ss << "std container and classes (c++14):\n";

    {
        std::integer_sequence<int, 9, 2, 5, 1, 9, 1, 6> value {};
        ss << formLabel( "std::integer_sequence" ) << doctest::toString( value ) << '\n';
    }
#endif

#if CXX_STANDARD >= 17
    ss << "std container and classes (c++17):\n";
    
    {
        const std::string s = "Hello world!";
        std::string_view value{ s };
        value = value.substr( 0, 5 );
        ss << formLabel( "std::string_view" ) << doctest::toString( value ) << '\n';
    }
    
    {
        ss << formLabel( "std::nullopt" ) << doctest::toString( std::nullopt ) << '\n';
        std::optional<int> value {};
        ss << formLabel( "std::optional(nullopt)" ) << doctest::toString( value ) << '\n';
        value = 9;
        ss << formLabel( "std::optional" ) << doctest::toString( value ) << '\n';
    }

    {
        std::variant<std::monostate, int, float> value {};
        ss << formLabel( "std::variant(monostate)" ) << doctest::toString( value ) << '\n';
        value = 5;
        ss << formLabel( "std::variant(int)" ) << doctest::toString( value ) << '\n';
        value = 9.2f;
        ss << formLabel( "std::variant(float)" ) << doctest::toString( value ) << '\n';
    }
#endif

#if CXX_STANDARD >= 20
    ss << "std container and classes (c++20):\n";
    
    {
        std::array<int, 3> ar { 9, 1, 5 };
        std::span<int> value { ar };
        ss << formLabel( "std::span" ) << doctest::toString( value ) << '\n';
    }
#endif
    
    
    Approvals::verify(ss.str());
}
