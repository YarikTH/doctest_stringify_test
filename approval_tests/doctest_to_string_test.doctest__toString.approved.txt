Fundamental types:
    std::nullptr_t:                 NULL
    short int:                      -1
    unsigned short int:             1
    int:                            -2
    unsigned int:                   2
    long int:                       -3
    unsigned long int:              3
    long long int:                  -4
    unsigned long long int:         4
    bool:                           true
    signed char:                    65
    unsigned char:                  66
    char:                           67
    wchar_t:                        {?}
    char16_t:                       {?}
    char32_t:                       {?}
    float:                          1.1f
    double:                         2.2
    long double:                    3.3
    short int*:                     NULL
    unsigned short int*:            0x0000000000000002
    int*:                           0x0000000000000003
    unsigned int*:                  0x0000000000000004
    long int*:                      0x0000000000000005
    unsigned long int*:             0x0000000000000006
    long long int*:                 0x0000000000000007
    unsigned long long int*:        0x0000000000000008
    bool*:                          0x0000000000000009
    signed char*:                   0x000000000000000a
    unsigned char*:                 0x000000000000000b
    char*:                          0x000000000000000c
    wchar_t*:                       0x000000000000000d
    char16_t*:                      0x000000000000000e
    char32_t*:                      0x000000000000000f
    float*:                         0x0000000000000010
    double*:                        0x0000000000000011
    long double*:                   0x0000000000000012
    const short int*:               NULL
    const unsigned short int*:      0x0000000000000002
    const int*:                     0x0000000000000003
    const unsigned int*:            0x0000000000000004
    const long int*:                0x0000000000000005
    const unsigned long int*:       0x0000000000000006
    const long long int*:           0x0000000000000007
    const unsigned long long int*:  0x0000000000000008
    const bool*:                    0x0000000000000009
    const signed char*:             0x000000000000000a
    const unsigned char*:           0x000000000000000b
    const char*:                    0x000000000000000c
    const wchar_t*:                 0x000000000000000d
    const char16_t*:                0x000000000000000e
    const char32_t*:                0x000000000000000f
    const float*:                   0x0000000000000010
    const double*:                  0x0000000000000011
    const long double*:             0x0000000000000012
ะก arrays:
    const char[]:                   foo
    char[]:                         bar
    const int[]:                    123
    int[]:                          456
enumerations:
    Enum:                           1
    EnumWithOss:                    2
    EnumClass:                      3
    EnumClassWithOss:               4
structs:
    struct:                         {?}
    struct with oss:                309
std container and classes (c++11):
    std::string:                    Hello world!
    std::pair:                      {?}
    std::tuple:                     {?}
    std::shared_ptr(NULL):          0
    std::shared_ptr:                0x17332c0
    std::unique_ptr(NULL):          0
    std::unique_ptr:                0x17332e0
    std::array:                     {?}
    std::valarray:                  {?}
    std::initializer_list:          {?}
    std::vector:                    {?}
    std::deque:                     {?}
    std::list:                      {?}
    std::forward_list:              {?}
    std::stack:                     {?}
    std::queue:                     {?}
    std::priority_queue:            {?}
    std::set:                       {?}
    std::multiset:                  {?}
    std::map:                       {?}
    std::multimap:                  {?}
std container and classes (c++14):
    std::integer_sequence:          {?}
std container and classes (c++17):
    std::string_view:               Hello
    std::nullopt:                   {?}
    std::optional(nullopt):         {?}
    std::optional:                  {?}
    std::variant(monostate):        {?}
    std::variant(int):              {?}
    std::variant(float):            {?}
std container and classes (c++20):
    std::span:                      {?}

